File: backend/routers/cohorts.py
import secrets
import random
from fastapi import APIRouter, Depends, HTTPException
from typing import List
from datetime import date, datetime

from database import get_convex
from schemas import (
    CohortCreate,
    CohortUpdate,
    CohortResponse,
    CohortDetailResponse,
    CohortMemberResponse,
    DailyChallengeCreate,
    DailyChallengeResponse,
)
from auth import get_required_user
from convex import ConvexClient

router = APIRouter(prefix="/api/cohorts", tags=["cohorts"])


def generate_invite_code(client: ConvexClient) -> str:
    """Generate a unique 6-character uppercase alphanumeric code."""
    while True:
        code = "".join(secrets.choice("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789") for _ in range(6))
        exists = client.query("cohorts:getByInviteCode", {"inviteCode": code})
        if not exists:
            return code


@router.post("/", response_model=CohortResponse, status_code=201)
def create_cohort(
    payload: CohortCreate,
    client: ConvexClient = Depends(get_convex),
    current_user: dict = Depends(get_required_user),
):
    """Create a new cohort coding league and generate a unique join invite code."""
    base_slug = payload.name.lower().strip().replace(" ", "-")
    base_slug = "".join(c for c in base_slug if c.isalnum() or c == "-")
    slug = base_slug
    counter = 1
    while client.query("cohorts:getBySlug", {"slug": slug}):
        slug = f"{base_slug}-{counter}"
        counter += 1

    invite_code = generate_invite_code(client)
    
    cohort_id = client.mutation("cohorts:create", {
        "name": payload.name,
        "slug": slug,
        "description": payload.description,
        "inviteCode": invite_code,
        "createdBy": current_user["id"]
    })
    
    res = client.query("cohorts:getBySlug", {"slug": slug})
    if res:
        res["id"] = str(res["_id"])
        res["created_by"] = res.get("createdBy")
        res["invite_code"] = res.get("inviteCode")
        res["created_at"] = res.get("_creationTime")
    return res


@router.post("/join", response_model=CohortResponse)
def join_cohort(
    payload: dict,
    client: ConvexClient = Depends(get_convex),
    current_user: dict = Depends(get_required_user),
):
    """Join a cohort coding league using a 6-character invite code."""
    invite_code = payload.get("invite_code")
    if not invite_code:
        raise HTTPException(status_code=400, detail="Invite code is required")
    
    invite_code = invite_code.strip().upper()
    cohort = client.query("cohorts:getByInviteCode", {"inviteCode": invite_code})
    if not cohort:
        raise HTTPException(status_code=404, detail="Cohort not found with this invite code")

    client.mutation("cohorts:join", {"cohortId": cohort.get("_id"), "userId": current_user["id"]})
    cohort["id"] = str(cohort["_id"])
    cohort["created_by"] = cohort.get("createdBy")
    cohort["invite_code"] = cohort.get("inviteCode")
    cohort["created_at"] = cohort.get("_creationTime")
    return cohort


@router.get("/me")
def get_my_cohorts(
    client: ConvexClient = Depends(get_convex),
    current_user: dict = Depends(get_required_user),
):
    """List all cohorts that the current user belongs to."""
    cohorts = client.query("cohorts:listForUser", {"userId": current_user["id"]})
    for c in cohorts:
        c["id"] = str(c.get("_id"))
        c["created_by"] = c.get("createdBy")
        c["invite_code"] = c.get("inviteCode")
        c["created_at"] = c.get("_creationTime")
    return cohorts


@router.get("/{slug}")
def get_cohort_detail(
    slug: str,
    client: ConvexClient = Depends(get_convex),
    current_user: dict = Depends(get_required_user),
):
    """Get detailed information about a cohort, including membership list."""
    cohort = client.query("cohorts:getBySlug", {"slug": slug})
    if not cohort:
        raise HTTPException(status_code=404, detail="Cohort not found")

    members = client.query("cohorts:getMembers", {"cohortId": cohort.get("_id")})
    is_member = any(m.get("userId") == current_user["id"] for m in members)
    
    if not is_member:
        raise HTTPException(status_code=403, detail="You are not a member of this cohort")

    return {
        "id": str(cohort.get("_id")),
        "name": cohort.get("name"),
        "slug": cohort.get("slug"),
        "description": cohort.get("description"),
        "invite_code": cohort.get("inviteCode"),
        "created_by": cohort.get("createdBy"),
        "created_at": cohort.get("_creationTime"),
        "members": [
            {
                "user_id": m.get("userId"),
                "username": m.get("username"),
                "display_name": m.get("displayName"),
                "avatar_url": m.get("avatarUrl"),
                "role": m.get("role"),
                "joined_at": m.get("joinedAt")
            } for m in members
        ]
    }


@router.post("/{slug}/leave", status_code=200)
def leave_cohort(
    slug: str,
    client: ConvexClient = Depends(get_convex),
    current_user: dict = Depends(get_required_user),
):
    """Leave a cohort."""
    cohort = client.query("cohorts:getBySlug", {"slug": slug})
    if not cohort:
        raise HTTPException(status_code=404, detail="Cohort not found")

    members = client.query("cohorts:getMembers", {"cohortId": cohort.get("_id")})
    user_member = next((m for m in members if m.get("userId") == current_user["id"]), None)
    
    if not user_member:
        raise HTTPException(status_code=400, detail="You are not a member of this cohort")

    if user_member.get("role") == "admin":
        admin_count = sum(1 for m in members if m.get("role") == "admin")
        if admin_count == 1:
            raise HTTPException(status_code=400, detail="You are the only admin. You cannot leave the cohort.")

    client.mutation("cohorts:leave", {"cohortId": cohort.get("_id"), "userId": current_user["id"]})
    return {"status": "success", "message": "Successfully left the cohort"}


@router.put("/{slug}")
def update_cohort(
    slug: str,
    payload: CohortUpdate,
    client: ConvexClient = Depends(get_convex),
    current_user: dict = Depends(get_required_user),
):
    """Update cohort details. Only admins can update."""
    cohort = client.query("cohorts:getBySlug", {"slug": slug})
    if not cohort:
        raise HTTPException(status_code=404, detail="Cohort not found")

    members = client.query("cohorts:getMembers", {"cohortId": cohort.get("_id")})
    user_member = next((m for m in members if m.get("userId") == current_user["id"]), None)
    
    if not user_member or user_member.get("role") != "admin":
        raise HTTPException(status_code=403, detail="Only admins can update the cohort")

    client.mutation("cohorts:update", {
        "cohortId": cohort.get("_id"),
        "name": payload.name,
        "description": payload.description
    })
    
    updated = client.query("cohorts:getBySlug", {"slug": slug})
    updated["id"] = str(updated["_id"])
    updated["created_by"] = updated.get("createdBy")
    updated["invite_code"] = updated.get("inviteCode")
    updated["created_at"] = updated.get("_creationTime")
    return updated


@router.delete("/{slug}", status_code=204)
def delete_cohort(
    slug: str,
    client: ConvexClient = Depends(get_convex),
    current_user: dict = Depends(get_required_user),
):
    """Delete a cohort entirely. Only admins can delete."""
    cohort = client.query("cohorts:getBySlug", {"slug": slug})
    if not cohort:
        raise HTTPException(status_code=404, detail="Cohort not found")

    members = client.query("cohorts:getMembers", {"cohortId": cohort.get("_id")})
    user_member = next((m for m in members if m.get("userId") == current_user["id"]), None)
    
    if not user_member or user_member.get("role") != "admin":
        raise HTTPException(status_code=403, detail="Only admins can delete the cohort")

    client.mutation("cohorts:deleteCohort", {"cohortId": cohort.get("_id")})
    return None


@router.get("/{slug}/daily-challenge")
def get_daily_challenge(
    slug: str,
    client: ConvexClient = Depends(get_convex),
    current_user: dict = Depends(get_required_user),
):
    """Get or generate the daily challenge for a cohort."""
    cohort = client.query("cohorts:getBySlug", {"slug": slug})
    if not cohort:
        raise HTTPException(status_code=404, detail="Cohort not found")

    members = client.query("cohorts:getMembers", {"cohortId": cohort.get("_id")})
    is_member = any(m.get("userId") == current_user["id"] for m in members)
    if not is_member:
        raise HTTPException(status_code=403, detail="Not a member of this cohort")

    today = date.today().isoformat()
    challenge = client.query("cohorts:getDailyChallenge", {"cohortId": cohort.get("_id"), "date": today})

    if not challenge:
        problems = client.query("problems:list", {})
        if not problems:
            raise HTTPException(status_code=404, detail="No problems available")
        problem = random.choice(problems)
        
        challenge_id = client.mutation("cohorts:createDailyChallenge", {
            "cohortId": cohort.get("_id"),
            "problemId": problem.get("_id"),
            "date": today
        })
        
        challenge = client.query("cohorts:getDailyChallenge", {"cohortId": cohort.get("_id"), "date": today})
    else:
        problem = client.query("problems:getById", {"problemId": challenge.get("problemId")})

    return {
        "id": str(challenge.get("_id")),
        "cohort_id": str(challenge.get("cohortId")),
        "problem_id": str(challenge.get("problemId")),
        "date": today,
        "problem_slug": problem.get("slug") if problem else None,
        "problem_title": problem.get("title") if problem else None
    }


@router.get("/{slug}/challenges/today")
def get_today_challenge(
    slug: str,
    client: ConvexClient = Depends(get_convex),
    current_user: dict = Depends(get_required_user),
):
    return get_daily_challenge(slug, client, current_user)

File: frontend/convex/cohorts.ts
import { mutation, query } from "./_generated/server";
import { v } from "convex/values";

export const create = mutation({
  args: {
    name: v.string(),
    slug: v.string(),
    description: v.optional(v.string()),
    inviteCode: v.string(),
    createdBy: v.string(),
  },
  handler: async (ctx, args) => {
    const cohortId = await ctx.db.insert("cohorts", {
      name: args.name,
      slug: args.slug,
      description: args.description,
      inviteCode: args.inviteCode,
      createdBy: args.createdBy,
    });

    await ctx.db.insert("cohortMembers", {
      cohortId,
      userId: args.createdBy,
      role: "admin",
    });

    return cohortId;
  },
});

export const getBySlug = query({
  args: { slug: v.string() },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("cohorts")
      .withIndex("by_slug", (q) => q.eq("slug", args.slug))
      .first();
  },
});

export const getByInviteCode = query({
  args: { inviteCode: v.string() },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("cohorts")
      .withIndex("by_inviteCode", (q) => q.eq("inviteCode", args.inviteCode))
      .first();
  },
});

export const join = mutation({
  args: { cohortId: v.id("cohorts"), userId: v.string() },
  handler: async (ctx, args) => {
    const existing = await ctx.db
      .query("cohortMembers")
      .withIndex("by_cohort", (q) => q.eq("cohortId", args.cohortId))
      .filter((q) => q.eq(q.field("userId"), args.userId))
      .first();

    if (!existing) {
      await ctx.db.insert("cohortMembers", {
        cohortId: args.cohortId,
        userId: args.userId,
        role: "member",
      });
    }
  },
});

export const listForUser = query({
  args: { userId: v.string() },
  handler: async (ctx, args) => {
    const memberships = await ctx.db
      .query("cohortMembers")
      .withIndex("by_user", (q) => q.eq("userId", args.userId))
      .collect();

    const cohorts = [];
    for (const m of memberships) {
      const cohort = await ctx.db.get(m.cohortId);
      if (cohort) {
        cohorts.push(cohort);
      }
    }
    return cohorts;
  },
});

export const createDailyChallenge = mutation({
  args: { cohortId: v.id("cohorts"), problemId: v.string(), date: v.string() },
  handler: async (ctx, args) => {
    return await ctx.db.insert("dailyChallenges", {
      cohortId: args.cohortId,
      problemId: args.problemId,
      date: args.date,
    });
  },
});

export const getDailyChallenge = query({
  args: { cohortId: v.id("cohorts"), date: v.string() },
  handler: async (ctx, args) => {
    return await ctx.db
      .query("dailyChallenges")
      .withIndex("by_cohort", (q) => q.eq("cohortId", args.cohortId))
      .filter((q) => q.eq(q.field("date"), args.date))
      .first();
  },
});

export const update = mutation({
  args: { cohortId: v.id("cohorts"), name: v.optional(v.string()), description: v.optional(v.string()) },
  handler: async (ctx, args) => {
    const patch: any = {};
    if (args.name !== undefined) patch.name = args.name;
    if (args.description !== undefined) patch.description = args.description;
    await ctx.db.patch(args.cohortId, patch);
  },
});

export const leave = mutation({
  args: { cohortId: v.id("cohorts"), userId: v.string() },
  handler: async (ctx, args) => {
    const member = await ctx.db
      .query("cohortMembers")
      .withIndex("by_cohort", (q) => q.eq("cohortId", args.cohortId))
      .filter((q) => q.eq(q.field("userId"), args.userId))
      .first();
    if (member) {
      await ctx.db.delete(member._id);
    }
  },
});

export const deleteCohort = mutation({
  args: { cohortId: v.id("cohorts") },
  handler: async (ctx, args) => {
    // delete members
    const members = await ctx.db.query("cohortMembers").withIndex("by_cohort", (q) => q.eq("cohortId", args.cohortId)).collect();
    for (const m of members) await ctx.db.delete(m._id);
    // delete challenges
    const challenges = await ctx.db.query("dailyChallenges").withIndex("by_cohort", (q) => q.eq("cohortId", args.cohortId)).collect();
    for (const c of challenges) await ctx.db.delete(c._id);
    
    await ctx.db.delete(args.cohortId);
  },
});

export const getMembers = query({
  args: { cohortId: v.id("cohorts") },
  handler: async (ctx, args) => {
    const members = await ctx.db.query("cohortMembers").withIndex("by_cohort", (q) => q.eq("cohortId", args.cohortId)).collect();
    const result = [];
    for (const m of members) {
      const u = await ctx.db.query("users").withIndex("by_userId", (q) => q.eq("userId", m.userId)).first();
      result.push({
        userId: m.userId,
        role: m.role,
        joinedAt: m._creationTime,
        username: u?.username,
        displayName: u?.displayName,
        avatarUrl: u?.avatarUrl
      });
    }
    return result;
  },
});

